lsm-perf.py

```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env
+
+import argparse
+import time
+import plumbum
+import sys
+import statistics
+from plumbum import local, SshMachine  # TODO: requirements
+
+
+ROUNDS = 3
+REPETITIONS_PER_ROUND = 50
+WARMUP_RUNS = 5
+ON_VM_WORKLOAD_PATH = '~/lsm-perf-workload'
+QEMU_EXIT_CMD = b'\x01cq\n'  # -> ctrl+a c q
+
+
+def main(args):
+    try:
+        init_output_file(args.out)
+        for round in range(ROUNDS):
+            print('Starting round %d' % round)
+            for kernel in args.kernels:
+                results = evaluating_kernel(
+                    kernel_path=kernel.name,
+                    img_path=args.image.name,
+                    workload_path=args.workload.name,
+                    keyfile=args.key.name
+                )
+                write_results_to_file(args.out, kernel.name, round, results)
+    except KeyboardInterrupt:
+        print('\nExit prematurely')
+    finally:
+        args.out.close()
+    return 0
+
+
+def evaluating_kernel(kernel_path, img_path, workload_path, keyfile):
+    """Starts a VM with the kernel and evaluates its
+    performances on the provided workload
+    """
+    results = []
+    name = kernel_path[kernel_path.rfind('/') + 1:]
+    print_eta(name, info='connecting')
+
+    with VM(kernel_path, img_path, keyfile) as vm:
+        vm.scp_to(workload_path, ON_VM_WORKLOAD_PATH)
+        work_cmd = vm.ssh[ON_VM_WORKLOAD_PATH]
+
+        print_eta(name, info='Running warm up')
+        for _ in range(WARMUP_RUNS):
+            work_cmd()
+
+        for i in range(REPETITIONS_PER_ROUND):
+            results.append(int(work_cmd().strip()))
+            percentage = (i + 1) * 100 / REPETITIONS_PER_ROUND
+            print_eta(name, info='%d%%' % percentage)
+
+        vm.ssh.path(ON_VM_WORKLOAD_PATH).delete()
+
+    stats = ('\taverage=%d, stdev=%d' %
+             (statistics.mean(results), statistics.stdev(results)))
+    print_eta(name, info=stats)
+    print()
+    return results
+
+
+class VM:
+    """Represents a qemu VM with an SSH connection.
+    To use with the `with` construct
+    """
+    def __init__(self, kernel_path, img_path, keyfile):
+        """Starts the qemu VM"""
+        qemu_args = construct_qemu_args(kernel_path, img_path)
+        self.process = local['qemu-system-x86_64'].popen(qemu_args)
+        self.ssh = None
+        self.key = keyfile
+
+    def __enter__(self):
+        """Initialize ssh connection"""
+        c = 5
+        while self.ssh is None:
+            time.sleep(1)
+            try:
+                self.ssh = SshMachine(
+                    '127.0.0.1', user='root', port=5555, keyfile=self.key)
+            except (EOFError, plumbum.machines.session.SSHCommsError) as e:
+                c -= 1
+                if c == 0:
+                    raise e
```

Comment:
I find the counter variable a little awkward. How about this instead?

```python
        for _ in range(5):
            time.sleep(1)
            try:
                self.ssh = SshMachine(
                    '127.0.0.1', user='root', port=5555, keyfile=self.key)
                break
            except (EOFError, plumbum.machines.session.SSHCommsError) as e:
                continue
```

Reply:
In this case the exception is not raised after many failures though. 
We would need 
```
        for _ in range(5):
            time.sleep(1)
            try:
                self.ssh = SshMachine(
                    '127.0.0.1', user='root', port=5555, keyfile=self.key)
                return self
            except (EOFError, plumbum.machines.session.SSHCommsError) as e:
                continue
        raise e
```
What do you think?

Reply:
Oops yeah you're right. You could do it that way, or try this: http://xion.io/post/code/python-retry-idiom.html

Reply:
Ok I changed it, and used the newly defined VMException